### PR TITLE
feat: implement OAuth login with Popup

### PR DESCRIPTION
**Pull request by sherzod-bakhodirov**
### 📦 Pull Request
https://magiclink.atlassian.net/browse/PDEEXP-1954

image

### ✅ Fixed Issues

- [List any fixed issues here like: Fixes #XXXX]

### 🚨 Test instructions

[login with Popup(video)](https://drive.google.com/file/d/18SYpH3fk3ITfzhA-d7jEKiXESSGkfqMs/view?usp=sharing)
[login with Popup and MFA(video)](https://drive.google.com/file/d/1fwGJmeR5LJ12nGwFK0igoIIvF4DlgHSX/view?usp=sharing)
[login with Popup and recovery flow(video)](https://drive.google.com/file/d/16UazReE_gyV2onYHS2ASY3nuRCbvRDRx/view?usp=sharing)

[Describe any additional context required to test the PR/feature/bug fix.]

### ⚠️ Don't forget to add a [semver](https://semver.org/) label! 
Please 🚨 **ONLY ADD ONE** 🚨 of the following labels, failing to do so may lead to adverse versioning of your changes when published:
- `patch`: Bug Fix?
- `minor`: New Feature?
- `major`: Breaking Change?
- `skip-release`: It's unnecessary to publish this change.

### Special Note
Please avoid adding any of the `Priority` labels as they conflict with the labels above ☝️

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @magic-ext/algorand@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/aptos@11.18.0-canary.828.11860318645.0
  npm install @magic-ext/avalanche@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/bitcoin@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/conflux@21.18.0-canary.828.11860318645.0
  npm install @magic-ext/cosmos@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/ed25519@19.18.0-canary.828.11860318645.0
  npm install @magic-ext/farcaster@0.18.0-canary.828.11860318645.0
  npm install @magic-ext/flow@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/gdkms@11.18.0-canary.828.11860318645.0
  npm install @magic-ext/harmony@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/icon@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/kadena@0.8.0-canary.828.11860318645.0
  npm install @magic-ext/near@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/oauth@22.18.0-canary.828.11860318645.0
  npm install @magic-ext/oauth2@9.18.0-canary.828.11860318645.0
  npm install @magic-ext/oidc@11.18.0-canary.828.11860318645.0
  npm install @magic-ext/polkadot@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/react-native-bare-oauth@25.19.0-canary.828.11860318645.0
  npm install @magic-ext/react-native-expo-oauth@25.19.0-canary.828.11860318645.0
  npm install @magic-ext/solana@25.19.0-canary.828.11860318645.0
  npm install @magic-ext/sui@0.19.0-canary.828.11860318645.0
  npm install @magic-ext/taquito@20.18.0-canary.828.11860318645.0
  npm install @magic-ext/terra@20.18.0-canary.828.11860318645.0
  npm install @magic-ext/tezos@23.18.0-canary.828.11860318645.0
  npm install @magic-ext/webauthn@22.18.0-canary.828.11860318645.0
  npm install @magic-ext/zilliqa@23.18.0-canary.828.11860318645.0
  npm install @magic-sdk/commons@24.18.0-canary.828.11860318645.0
  npm install @magic-sdk/pnp@22.19.0-canary.828.11860318645.0
  npm install @magic-sdk/provider@28.18.0-canary.828.11860318645.0
  npm install @magic-sdk/react-native-bare@29.19.0-canary.828.11860318645.0
  npm install @magic-sdk/react-native-expo@29.19.0-canary.828.11860318645.0
  npm install @magic-sdk/types@24.16.0-canary.828.11860318645.0
  npm install magic-sdk@28.19.0-canary.828.11860318645.0
  # or 
  yarn add @magic-ext/algorand@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/aptos@11.18.0-canary.828.11860318645.0
  yarn add @magic-ext/avalanche@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/bitcoin@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/conflux@21.18.0-canary.828.11860318645.0
  yarn add @magic-ext/cosmos@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/ed25519@19.18.0-canary.828.11860318645.0
  yarn add @magic-ext/farcaster@0.18.0-canary.828.11860318645.0
  yarn add @magic-ext/flow@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/gdkms@11.18.0-canary.828.11860318645.0
  yarn add @magic-ext/harmony@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/icon@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/kadena@0.8.0-canary.828.11860318645.0
  yarn add @magic-ext/near@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/oauth@22.18.0-canary.828.11860318645.0
  yarn add @magic-ext/oauth2@9.18.0-canary.828.11860318645.0
  yarn add @magic-ext/oidc@11.18.0-canary.828.11860318645.0
  yarn add @magic-ext/polkadot@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/react-native-bare-oauth@25.19.0-canary.828.11860318645.0
  yarn add @magic-ext/react-native-expo-oauth@25.19.0-canary.828.11860318645.0
  yarn add @magic-ext/solana@25.19.0-canary.828.11860318645.0
  yarn add @magic-ext/sui@0.19.0-canary.828.11860318645.0
  yarn add @magic-ext/taquito@20.18.0-canary.828.11860318645.0
  yarn add @magic-ext/terra@20.18.0-canary.828.11860318645.0
  yarn add @magic-ext/tezos@23.18.0-canary.828.11860318645.0
  yarn add @magic-ext/webauthn@22.18.0-canary.828.11860318645.0
  yarn add @magic-ext/zilliqa@23.18.0-canary.828.11860318645.0
  yarn add @magic-sdk/commons@24.18.0-canary.828.11860318645.0
  yarn add @magic-sdk/pnp@22.19.0-canary.828.11860318645.0
  yarn add @magic-sdk/provider@28.18.0-canary.828.11860318645.0
  yarn add @magic-sdk/react-native-bare@29.19.0-canary.828.11860318645.0
  yarn add @magic-sdk/react-native-expo@29.19.0-canary.828.11860318645.0
  yarn add @magic-sdk/types@24.16.0-canary.828.11860318645.0
  yarn add magic-sdk@28.19.0-canary.828.11860318645.0
  ```
</details>

### CODE REVIEW

File: packages/@magic-ext/oauth2/src/index.ts [hunk 1]
```diff
@@ -64,6 +65,90 @@ export class OAuthExtension extends Extension.Internal<'oauth2'> {
...
+      // Reject if popup already exists to prevent multiple instances
+      if (popup) {
+        reject(new Error('Popup window already exists'));
+        return;
+      }
+
+      const parseRedirectResult = this.utils.createJsonRpcRequestPayload(OAuthPayloadMethods.Start, [
+        {
+          ...configuration,
+          redirectURI: 'http://localhost:3024/oauth2/popup/verify',
```
Comment (Ethella) [inline packages/@magic-ext/oauth2/src/index.ts:86]: Do we still need this redirectURI?

Comment (sherzod-bakhodirov) [reply packages/@magic-ext/oauth2/src/index.ts:86]: yes, this URI will be loaded inside of child window to send back the success params

Comment (sherzod-bakhodirov) [reply packages/@magic-ext/oauth2/src/index.ts]: Jerry I have changed the implementation according to how we discussed it, now I am passing an extra property - `loginWithPopup` and then assigning redirectURI in iframe i.e. mandrake in main window

**Review by Ethella — APPROVED** [review]

File: packages/@magic-ext/oauth2/src/index.ts [hunk 2]
```diff
@@ -64,6 +65,90 @@ export class OAuthExtension extends Extension.Internal<'oauth2'> {
...
+    const top = window.screenTop + window.outerHeight * 0.15;
+
+    return this.utils.createPromiEvent<OAuthRedirectResult>(async (resolve, reject) => {
+      // Reject if popup already exists to prevent multiple instances
+      if (popup) {
+        reject(new Error('Popup window already exists'));
+        return;
+      }
+
+      const parseRedirectResult = this.utils.createJsonRpcRequestPayload(OAuthPayloadMethods.Start, [
```
Comment (Ethella) [inline packages/@magic-ext/oauth2/src/index.ts:83]: Create a popup RPC method

**Review by Ethella — DISMISSED** [review]

**Review by Ethella — APPROVED** [review]

**Pull request merged**